In [2]:
import pandas as pd
import numpy as np

Datasets:

In [3]:
deliveries = pd.read_csv('deliveries.csv')
matches = pd.read_csv('matches.csv')

In [4]:
deliveries.head()

match_id  inning         batting_team                 bowling_team  over  \
0         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
1         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
2         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
3         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
4         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   

   ball    batsman non_striker    bowler  is_super_over  ...  bye_runs  \
0     1  DA Warner    S Dhawan  TS Mills              0  ...         0   
1     2  DA Warner    S Dhawan  TS Mills              0  ...         0   
2     3  DA Warner    S Dhawan  TS Mills              0  ...         0   
3     4  DA Warner    S Dhawan  TS Mills              0  ...         0   
4     5  DA Warner    S Dhawan  TS Mills              0  ...         0   

   legbye_runs  noball_runs  penalty_runs  batsman_runs  extra_runs  \
0            0            0             0             0           0   
1            0            0             0             0           0   
2            0            0             0             4           0   
3            0            0             0             0           0   
4            0            0             0             0           2   

   total_runs  player_dismissed dismissal_kind fielder  
0           0               NaN            NaN     NaN  
1           0               NaN            NaN     NaN  
2           4               NaN            NaN     NaN  
3           0               NaN            NaN     NaN  
4           2               NaN            NaN     NaN  

[5 rows x 21 columns]

In [5]:
deliveries.shape

(179078, 21)

This signifies that there have been 1,79,078 deliveries in the ipl franchise

In [6]:
deliveries.columns

Index(['match_id', 'inning', 'batting_team', 'bowling_team', 'over', 'ball',
       'batsman', 'non_striker', 'bowler', 'is_super_over', 'wide_runs',
       'bye_runs', 'legbye_runs', 'noball_runs', 'penalty_runs',
       'batsman_runs', 'extra_runs', 'total_runs', 'player_dismissed',
       'dismissal_kind', 'fielder'],
      dtype='object')

Missing values

In [7]:
missing_values = deliveries.isnull().sum().reset_index().assign(
 missing_percentage = lambda x: (x[0]/deliveries.shape[0])*100  
).sort_values(by="missing_percentage",ascending=False).rename(columns={
    "index":"column name", 0:"missing row count","missing_percentage":"missing percentage"
})
missing_values = missing_values[missing_values['missing percentage']>0]
missing_values

column name  missing row count  missing percentage
20           fielder             172630           96.399334
19    dismissal_kind             170244           95.066954
18  player_dismissed             170244           95.066954

Fielder, dismissal kind and player dismissed has so many missing values. They are bound to have missing values because not every ball records dismissal kind and player dismissed 

In [8]:
deliveries.nunique()

match_id            756
inning                5
batting_team         15
bowling_team         15
over                 20
ball                  9
batsman             516
non_striker         511
bowler              405
is_super_over         2
wide_runs             6
bye_runs              5
legbye_runs           6
noball_runs           5
penalty_runs          2
batsman_runs          8
extra_runs            7
total_runs           10
player_dismissed    487
dismissal_kind        9
fielder             499
dtype: int64

Match_id 756 indicates that the data consists of 756 matches

In [9]:
deliveries.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 179078 entries, 0 to 179077
Data columns (total 21 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   match_id          179078 non-null  int64 
 1   inning            179078 non-null  int64 
 2   batting_team      179078 non-null  object
 3   bowling_team      179078 non-null  object
 4   over              179078 non-null  int64 
 5   ball              179078 non-null  int64 
 6   batsman           179078 non-null  object
 7   non_striker       179078 non-null  object
 8   bowler            179078 non-null  object
 9   is_super_over     179078 non-null  int64 
 10  wide_runs         179078 non-null  int64 
 11  bye_runs          179078 non-null  int64 
 12  legbye_runs       179078 non-null  int64 
 13  noball_runs       179078 non-null  int64 
 14  penalty_runs      179078 non-null  int64 
 15  batsman_runs      179078 non-null  int64 
 16  extra_runs        179078 non-null  int

In [10]:
matches.head()

id  season       city        date                        team1  \
0   1    2017  Hyderabad  2017-04-05          Sunrisers Hyderabad   
1   2    2017       Pune  2017-04-06               Mumbai Indians   
2   3    2017     Rajkot  2017-04-07                Gujarat Lions   
3   4    2017     Indore  2017-04-08       Rising Pune Supergiant   
4   5    2017  Bangalore  2017-04-08  Royal Challengers Bangalore   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1       Rising Pune Supergiant       Rising Pune Supergiant         field   
2        Kolkata Knight Riders        Kolkata Knight Riders         field   
3              Kings XI Punjab              Kings XI Punjab         field   
4             Delhi Daredevils  Royal Challengers Bangalore           bat   

   result  dl_applied                       winner  win_by_runs  \
0  normal           0          Sunrisers Hyderabad           35   
1  normal           0       Rising Pune Supergiant            0   
2  normal           0        Kolkata Knight Riders            0   
3  normal           0              Kings XI Punjab            0   
4  normal           0  Royal Challengers Bangalore           15   

   win_by_wickets player_of_match                                      venue  \
0               0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
1               7       SPD Smith    Maharashtra Cricket Association Stadium   
2              10         CA Lynn     Saurashtra Cricket Association Stadium   
3               6      GJ Maxwell                     Holkar Cricket Stadium   
4               0       KM Jadhav                      M Chinnaswamy Stadium   

          umpire1        umpire2 umpire3  
0     AY Dandekar       NJ Llong     NaN  
1  A Nand Kishore         S Ravi     NaN  
2     Nitin Menon      CK Nandan     NaN  
3    AK Chaudhary  C Shamshuddin     NaN  
4             NaN            NaN     NaN

In [11]:
matches.shape

(756, 18)

756 rows indicates that there are 756 matches. It matches with the deliveries dataset as well. This implies that there is no inconsistency in the data.

In [12]:
matches.columns

Index(['id', 'season', 'city', 'date', 'team1', 'team2', 'toss_winner',
       'toss_decision', 'result', 'dl_applied', 'winner', 'win_by_runs',
       'win_by_wickets', 'player_of_match', 'venue', 'umpire1', 'umpire2',
       'umpire3'],
      dtype='object')

In [13]:
missing_values = matches.isnull().sum().reset_index().assign(
 missing_percentage = lambda x: (x[0]/matches.shape[0])*100  
).sort_values(by="missing_percentage",ascending=False).rename(columns={
    "index":"column name", 0:"missing row count","missing_percentage":"missing percentage"
})
missing_values = missing_values[missing_values['missing percentage']>0]
missing_values

column name  missing row count  missing percentage
17          umpire3                637           84.259259
2              city                  7            0.925926
10           winner                  4            0.529101
13  player_of_match                  4            0.529101
16          umpire2                  2            0.264550
15          umpire1                  2            0.264550

Percentage of missing values for the columns city, winner, player_of_match, umpire2, umpire1 is less than 10%, it would affect the data significantly. Umpire 3 is present only when the team captain asks for a review.

In [14]:
matches.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 756 entries, 0 to 755
Data columns (total 18 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   id               756 non-null    int64 
 1   season           756 non-null    int64 
 2   city             749 non-null    object
 3   date             756 non-null    object
 4   team1            756 non-null    object
 5   team2            756 non-null    object
 6   toss_winner      756 non-null    object
 7   toss_decision    756 non-null    object
 8   result           756 non-null    object
 9   dl_applied       756 non-null    int64 
 10  winner           752 non-null    object
 11  win_by_runs      756 non-null    int64 
 12  win_by_wickets   756 non-null    int64 
 13  player_of_match  752 non-null    object
 14  venue            756 non-null    object
 15  umpire1          754 non-null    object
 16  umpire2          754 non-null    object
 17  umpire3          119 non-null    ob

In [15]:
matches['season'].unique()

array([2017, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2018,
       2019], dtype=int64)

The data is recorded from matches that have taken place between 2008-2019 i.e., 12 seasons

In [16]:
matches['team1'].unique()

array(['Sunrisers Hyderabad', 'Mumbai Indians', 'Gujarat Lions',
       'Rising Pune Supergiant', 'Royal Challengers Bangalore',
       'Kolkata Knight Riders', 'Delhi Daredevils', 'Kings XI Punjab',
       'Chennai Super Kings', 'Rajasthan Royals', 'Deccan Chargers',
       'Kochi Tuskers Kerala', 'Pune Warriors', 'Rising Pune Supergiants',
       'Delhi Capitals'], dtype=object)

In [17]:
matches['team2'].unique()

array(['Royal Challengers Bangalore', 'Rising Pune Supergiant',
       'Kolkata Knight Riders', 'Kings XI Punjab', 'Delhi Daredevils',
       'Sunrisers Hyderabad', 'Mumbai Indians', 'Gujarat Lions',
       'Rajasthan Royals', 'Chennai Super Kings', 'Deccan Chargers',
       'Pune Warriors', 'Kochi Tuskers Kerala', 'Rising Pune Supergiants',
       'Delhi Capitals'], dtype=object)

Playing teams include SRH,RCB,MI,GL,DD,KXIP,CSK,RPS,RR,DC,KTK,Deccan Chargers.

Even though Delhi Capitals and Delhi Daredevils, Pune Warrior and Rising Pune Supergiant are coming from the same cities, Delhi and Pune, they are two different teams

Rising Pune Supergiant and Rising Pune Supergiants are the same team. So, ideally, we replace Rising Pune Supergiants to Rising Pune Supergiant.

In [18]:
matches['team1'] = matches['team1'].replace(to_replace=['Rising Pune Supergiants'],value='Rising Pune Supergiant')
matches['team2'] = matches['team2'].replace(to_replace=['Rising Pune Supergiants'],value='Rising Pune Supergiant')
matches['toss_winner'] = matches['toss_winner'].replace(to_replace=['Rising Pune Supergiants'],value='Rising Pune Supergiant')
matches['winner'] = matches['winner'].replace(to_replace=['Rising Pune Supergiants'],value='Rising Pune Supergiant')

In [19]:
deliveries['batting_team'] = deliveries['batting_team'].replace(to_replace=['Rising Pune Supergiants'],value='Rising Pune Supergiant')
deliveries['bowling_team'] = deliveries['bowling_team'].replace(to_replace=['Rising Pune Supergiants'],value='Rising Pune Supergiant')

In [20]:
PW = matches[matches['team1'] == 'Pune Warriors']
PW['season'].unique()

array([2011, 2012, 2013], dtype=int64)

Pune Warriors were a part of ipl from 2011-2013

In [21]:
RPS = matches[matches['team1'] == 'Rising Pune Supergiant']
RPS['season'].unique()

array([2017, 2016], dtype=int64)

Rising Pune Supergiant played ipl in 2016 and 2017

In [22]:
DD = matches[matches['team1'] == 'Delhi Daredevils']
DD['season'].unique()

array([2017, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2018],
      dtype=int64)

Delhi Daredevils were in ipl from 2008-2018

In [23]:
DC = matches[matches['team1'] == 'Delhi Capitals']
DC['season'].unique()

array([2019], dtype=int64)

Delhi Capitals started playing from 2019 

In [24]:
deliveries['inning'].unique()

array([1, 2, 3, 4, 5], dtype=int64)

There are upto 5 innings being played

In [25]:
deliveries['is_super_over'].sum()

81

There have been 81 superovers taken place from 2008-2019

In [26]:
matches['city'].unique()

array(['Hyderabad', 'Pune', 'Rajkot', 'Indore', 'Bangalore', 'Mumbai',
       'Kolkata', 'Delhi', 'Chandigarh', 'Kanpur', 'Jaipur', 'Chennai',
       'Cape Town', 'Port Elizabeth', 'Durban', 'Centurion',
       'East London', 'Johannesburg', 'Kimberley', 'Bloemfontein',
       'Ahmedabad', 'Cuttack', 'Nagpur', 'Dharamsala', 'Kochi',
       'Visakhapatnam', 'Raipur', 'Ranchi', 'Abu Dhabi', 'Sharjah', nan,
       'Mohali', 'Bengaluru'], dtype=object)

Matches have taken place in these cities

In [27]:
match_percent = matches['city'].value_counts().rename("Freq").reset_index().assign(
percentage = lambda x:(x['Freq']/matches.shape[0])*100
)
match_percent

index  Freq  percentage
0           Mumbai   101   13.359788
1          Kolkata    77   10.185185
2            Delhi    74    9.788360
3        Bangalore    66    8.730159
4        Hyderabad    64    8.465608
5          Chennai    57    7.539683
6           Jaipur    47    6.216931
7       Chandigarh    46    6.084656
8             Pune    38    5.026455
9           Durban    15    1.984127
10       Bengaluru    14    1.851852
11   Visakhapatnam    13    1.719577
12       Centurion    12    1.587302
13       Ahmedabad    12    1.587302
14          Rajkot    10    1.322751
15          Mohali    10    1.322751
16          Indore     9    1.190476
17      Dharamsala     9    1.190476
18    Johannesburg     8    1.058201
19         Cuttack     7    0.925926
20          Ranchi     7    0.925926
21  Port Elizabeth     7    0.925926
22       Cape Town     7    0.925926
23       Abu Dhabi     7    0.925926
24         Sharjah     6    0.793651
25          Raipur     6    0.793651
26           Kochi     5    0.661376
27          Kanpur     4    0.529101
28          Nagpur     3    0.396825
29       Kimberley     3    0.396825
30     East London     3    0.396825
31    Bloemfontein     2    0.264550

Most matches have taken place in Mumbai

In [28]:
match_percent.to_excel('C:/Users/Alexei/Desktop/Sparks Internship/Sports EDA.xlsx',sheet_name = 'Sheet1')

In [29]:
venue = matches['venue'].value_counts().rename("Freq").reset_index().assign(
percentage = lambda x:(x['Freq']/matches.shape[0])*100
)
venue

index  Freq  percentage
0                                        Eden Gardens    77   10.185185
1                               M Chinnaswamy Stadium    73    9.656085
2                                    Wankhede Stadium    73    9.656085
3                                    Feroz Shah Kotla    67    8.862434
4           Rajiv Gandhi International Stadium, Uppal    56    7.407407
5                     MA Chidambaram Stadium, Chepauk    49    6.481481
6                              Sawai Mansingh Stadium    47    6.216931
7          Punjab Cricket Association Stadium, Mohali    35    4.629630
8             Maharashtra Cricket Association Stadium    21    2.777778
9                          Subrata Roy Sahara Stadium    17    2.248677
10                         Dr DY Patil Sports Academy    17    2.248677
11                                          Kingsmead    15    1.984127
12  Punjab Cricket Association IS Bindra Stadium, ...    14    1.851852
13                                    SuperSport Park    12    1.587302
14                       Sardar Patel Stadium, Motera    12    1.587302
15  Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket St...    11    1.455026
16                                  Brabourne Stadium    11    1.455026
17             Saurashtra Cricket Association Stadium    10    1.322751
18                             Holkar Cricket Stadium     9    1.190476
19       Himachal Pradesh Cricket Association Stadium     9    1.190476
20                 Rajiv Gandhi Intl. Cricket Stadium     8    1.058201
21                          M. A. Chidambaram Stadium     8    1.058201
22                              New Wanderers Stadium     8    1.058201
23                            Feroz Shah Kotla Ground     7    0.925926
24                                   Barabati Stadium     7    0.925926
25                             M. Chinnaswamy Stadium     7    0.925926
26                                   St George's Park     7    0.925926
27                                           Newlands     7    0.925926
28                 JSCA International Stadium Complex     7    0.925926
29                               Sheikh Zayed Stadium     7    0.925926
30                Dubai International Cricket Stadium     7    0.925926
31                                  IS Bindra Stadium     7    0.925926
32   Shaheed Veer Narayan Singh International Stadium     6    0.793651
33                            Sharjah Cricket Stadium     6    0.793651
34                                      Nehru Stadium     5    0.661376
35                                         Green Park     4    0.529101
36                              De Beers Diamond Oval     3    0.396825
37       Vidarbha Cricket Association Stadium, Jamtha     3    0.396825
38                                       Buffalo Park     3    0.396825
39                                    OUTsurance Oval     2    0.264550
40                                   ACA-VDCA Stadium     2    0.264550

In [30]:
venue.to_excel('C:/Users/Alexei/Desktop/Sparks Internship/Sports EDA.xlsx',sheet_name = 'Sheet2')

Eden Gardens, M Chinnaswamy and Wankhede Stadium conducted the most number of matches

Imported to excel for visualizations

# Team Level Statistics

In [31]:
Toss_winner = matches['toss_winner'].value_counts().rename("Freq").reset_index()
Toss_winner

index  Freq
0                Mumbai Indians    98
1         Kolkata Knight Riders    92
2           Chennai Super Kings    89
3   Royal Challengers Bangalore    81
4               Kings XI Punjab    81
5              Delhi Daredevils    80
6              Rajasthan Royals    80
7           Sunrisers Hyderabad    46
8               Deccan Chargers    43
9                 Pune Warriors    20
10                Gujarat Lions    15
11       Rising Pune Supergiant    13
12               Delhi Capitals    10
13         Kochi Tuskers Kerala     8

Mumbai Indians have won the most number of tosses while Kochi Tuskers Kerela have lost the most

In [32]:
Toss_winner.to_excel('C:/Users/Alexei/Desktop/Sparks Internship/Sports EDA.xlsx',sheet_name = 'Sheet1')

In [33]:
match_winner = matches['winner'].value_counts().rename("Winner").reset_index()
match_winner

index  Winner
0                Mumbai Indians     109
1           Chennai Super Kings     100
2         Kolkata Knight Riders      92
3   Royal Challengers Bangalore      84
4               Kings XI Punjab      82
5              Rajasthan Royals      75
6              Delhi Daredevils      67
7           Sunrisers Hyderabad      58
8               Deccan Chargers      29
9        Rising Pune Supergiant      15
10                Gujarat Lions      13
11                Pune Warriors      12
12               Delhi Capitals      10
13         Kochi Tuskers Kerala       6

In [34]:
match_winner.to_excel('C:/Users/Alexei/Desktop/Sparks Internship/Sports EDA.xlsx',sheet_name = 'Sheet1')

Mumbai Indians have won the most number of matches as well

#  Chi-Square test of independence

To determine if the team winning the toss implies winning the match

In [35]:
crosstab = pd.crosstab(Toss_winner['Freq'],match_winner['Winner'])
crosstab

Winner  6    10   12   13   15   29   58   67   75   82   84   92   100  109
Freq                                                                        
8         1    0    0    0    0    0    0    0    0    0    0    0    0    0
10        0    1    0    0    0    0    0    0    0    0    0    0    0    0
13        0    0    1    0    0    0    0    0    0    0    0    0    0    0
15        0    0    0    1    0    0    0    0    0    0    0    0    0    0
20        0    0    0    0    1    0    0    0    0    0    0    0    0    0
43        0    0    0    0    0    1    0    0    0    0    0    0    0    0
46        0    0    0    0    0    0    1    0    0    0    0    0    0    0
80        0    0    0    0    0    0    0    1    1    0    0    0    0    0
81        0    0    0    0    0    0    0    0    0    1    1    0    0    0
89        0    0    0    0    0    0    0    0    0    0    0    1    0    0
92        0    0    0    0    0    0    0    0    0    0    0    0    1    0
98        0    0    0    0    0    0    0    0    0    0    0    0    0    1

In [36]:
import scipy.stats as stats
stats.chi2_contingency(crosstab)

(154.00000000000003,
 0.25026258025232334,
 143,
 array([[0.07142857, 0.07142857, 0.07142857, 0.07142857, 0.07142857,
         0.07142857, 0.07142857, 0.07142857, 0.07142857, 0.07142857,
         0.07142857, 0.07142857, 0.07142857, 0.07142857],
        [0.07142857, 0.07142857, 0.07142857, 0.07142857, 0.07142857,
         0.07142857, 0.07142857, 0.07142857, 0.07142857, 0.07142857,
         0.07142857, 0.07142857, 0.07142857, 0.07142857],
        [0.07142857, 0.07142857, 0.07142857, 0.07142857, 0.07142857,
         0.07142857, 0.07142857, 0.07142857, 0.07142857, 0.07142857,
         0.07142857, 0.07142857, 0.07142857, 0.07142857],
        [0.07142857, 0.07142857, 0.07142857, 0.07142857, 0.07142857,
         0.07142857, 0.07142857, 0.07142857, 0.07142857, 0.07142857,
         0.07142857, 0.07142857, 0.07142857, 0.07142857],
        [0.07142857, 0.07142857, 0.07142857, 0.07142857, 0.07142857,
         0.07142857, 0.07142857, 0.07142857, 0.07142857, 0.07142857,
         0.07142857, 0.071428

In [37]:
toss_decision = matches['toss_decision'].value_counts().rename("Freq").reset_index().assign(
percentage = lambda x:(x['Freq']/matches.shape[0])*100
)
toss_decision

index  Freq  percentage
0  field   463   61.243386
1    bat   293   38.756614

Ideally, a captain can either choose to field or bat

In [38]:
matches_winner_toss = matches.groupby(['winner','toss_winner'])['id'].size().reset_index()
matches_winner_toss

winner                  toss_winner  id
0    Chennai Super Kings          Chennai Super Kings  57
1    Chennai Super Kings              Deccan Chargers   2
2    Chennai Super Kings               Delhi Capitals   2
3    Chennai Super Kings             Delhi Daredevils   5
4    Chennai Super Kings              Kings XI Punjab   3
..                   ...                          ...  ..
113  Sunrisers Hyderabad               Mumbai Indians   5
114  Sunrisers Hyderabad                Pune Warriors   2
115  Sunrisers Hyderabad             Rajasthan Royals   3
116  Sunrisers Hyderabad  Royal Challengers Bangalore   6
117  Sunrisers Hyderabad          Sunrisers Hyderabad  23

[118 rows x 3 columns]

In [39]:
matches_winner_toss.to_excel('C:/Users/Alexei/Desktop/Sparks Internship/Sports EDA 2.xlsx',sheet_name = 'Sheet1')

To compute the count of matches won when a team won the toss

In [40]:
toss_decision.to_excel('C:/Users/Alexei/Desktop/Sparks Internship/Sports EDA.xlsx',sheet_name = 'Sheet1')

In [41]:
matches_team1 = matches['team1'].value_counts()
matches_team1

Mumbai Indians                 101
Kings XI Punjab                 91
Chennai Super Kings             89
Royal Challengers Bangalore     85
Kolkata Knight Riders           83
Delhi Daredevils                72
Rajasthan Royals                67
Sunrisers Hyderabad             63
Deccan Chargers                 43
Pune Warriors                   20
Rising Pune Supergiant          15
Gujarat Lions                   14
Kochi Tuskers Kerala             7
Delhi Capitals                   6
Name: team1, dtype: int64

In [42]:
matches_team2 = matches['team2'].value_counts()
matches_team2

Royal Challengers Bangalore    95
Kolkata Knight Riders          95
Delhi Daredevils               89
Mumbai Indians                 86
Kings XI Punjab                85
Rajasthan Royals               80
Chennai Super Kings            75
Sunrisers Hyderabad            45
Deccan Chargers                32
Pune Warriors                  26
Gujarat Lions                  16
Rising Pune Supergiant         15
Delhi Capitals                 10
Kochi Tuskers Kerala            7
Name: team2, dtype: int64

Combining these values we'll get the count of total matches played per team

In [43]:
runs_per_team = deliveries.groupby(['batting_team'])['total_runs'].sum().sort_values(ascending=False)
runs_per_team

batting_team
Mumbai Indians                 29809
Royal Challengers Bangalore    28126
Kings XI Punjab                27893
Kolkata Knight Riders          27419
Chennai Super Kings            26418
Delhi Daredevils               24388
Rajasthan Royals               22431
Sunrisers Hyderabad            17059
Deccan Chargers                11463
Pune Warriors                   6358
Gujarat Lions                   4862
Rising Pune Supergiant          4533
Delhi Capitals                  2630
Kochi Tuskers Kerala            1901
Name: total_runs, dtype: int64

In [44]:
runs_per_team.to_excel('C:/Users/Alexei/Desktop/Sparks Internship/Sports EDA.xlsx',sheet_name = 'Sheet1')

Mumbai indians scored the highest number of runs, followed by Royal Challengers Bangalore

In [45]:
runs_per_match = deliveries.groupby(['match_id','batting_team'])['total_runs'].sum().reset_index()
runs_per_match

match_id                 batting_team  total_runs
0            1  Royal Challengers Bangalore         172
1            1          Sunrisers Hyderabad         207
2            2               Mumbai Indians         184
3            2       Rising Pune Supergiant         187
4            3                Gujarat Lions         183
...        ...                          ...         ...
1505     11413          Sunrisers Hyderabad         171
1506     11414          Chennai Super Kings         162
1507     11414               Delhi Capitals         155
1508     11415          Chennai Super Kings         157
1509     11415               Mumbai Indians         152

[1510 rows x 3 columns]

In [46]:
avg_runs = runs_per_match.groupby(['batting_team'])['total_runs'].mean().reset_index()
avg_runs

batting_team  total_runs
0           Chennai Super Kings  161.085366
1               Deccan Chargers  152.840000
2                Delhi Capitals  164.375000
3              Delhi Daredevils  151.478261
4                 Gujarat Lions  162.066667
5               Kings XI Punjab  158.482955
6          Kochi Tuskers Kerala  135.785714
7         Kolkata Knight Riders  154.039326
8                Mumbai Indians  159.406417
9                 Pune Warriors  141.288889
10             Rajasthan Royals  153.636986
11       Rising Pune Supergiant  151.100000
12  Royal Challengers Bangalore  156.255556
13          Sunrisers Hyderabad  157.953704

Chennai Super Kings usually score the highest average runs in a match 

In [47]:
avg_runs.to_excel('C:/Users/Alexei/Desktop/Sparks Internship/Sports EDA.xlsx',sheet_name = 'Sheet1')

In [48]:
highest_total = deliveries.groupby(['match_id','batting_team']).sum()['total_runs'].sort_values(ascending=False)[:10]
highest_total 

match_id  batting_team               
411       Royal Challengers Bangalore    263
7937      Kolkata Knight Riders          250
620       Royal Challengers Bangalore    248
206       Chennai Super Kings            246
11338     Kolkata Knight Riders          241
61        Chennai Super Kings            240
562       Royal Challengers Bangalore    235
11147     Sunrisers Hyderabad            233
296       Kings XI Punjab                232
11339     Sunrisers Hyderabad            232
Name: total_runs, dtype: int64

Royal Challengers Bangalore have made the highest total ever in the IPL franchise

In [49]:
result = matches['result'].value_counts().rename("Freq").reset_index().assign(
percentage = lambda x:(x['Freq']/matches.shape[0])*100
)
result

index  Freq  percentage
0     normal   743   98.280423
1        tie     9    1.190476
2  no result     4    0.529101

The results have either been normal or tie and occasionally no result.

# Batsman Statistics

In [50]:
deliveries['Cumulative_batsman_runs'] = deliveries.groupby(['match_id','batsman'])['batsman_runs'].cumsum()
deliveries['ballcount'] = 1
deliveries['cummulative_ball_count'] = deliveries.groupby(['match_id','batsman'])['ballcount'].cumsum()
deliveries['Fours'] = deliveries.batsman_runs.apply(lambda x:1 if x==4 else 0)
deliveries['Sixes'] = deliveries.batsman_runs.apply(lambda x:1 if x==6 else 0)
deliveries['No Ball'] = deliveries.noball_runs.apply(lambda x:1 if x>0 else  0)
deliveries['Wide Ball'] = deliveries.wide_runs.apply(lambda x:1 if x>0 else 0)
deliveries['Maiden Overs'] = deliveries.total_runs.apply(lambda x:1 if x==0 else 0)
deliveries['Bounds'] = deliveries.total_runs.apply(lambda x: 1 if x==4 or x==6 else 0)
deliveries['Matches_played'] = deliveries.groupby(['batsman'])['match_id'].cumsum()

In [51]:
batsman_stats = (pd.concat([
    deliveries.batsman.value_counts().rename("Total Balls Faced"),
    deliveries.non_striker.value_counts().rename("Balls at Non-Stricker's End"),
    deliveries.groupby(["batsman"])["batsman_runs"].sum().rename("Total Runs Scored"),
    deliveries.groupby(["batsman"])["Cumulative_batsman_runs"].max().rename("Maximum Runs in a Match"),
    deliveries.groupby(["batsman"])["Fours"].sum().rename("Fours"),
    deliveries.groupby(["batsman"])["Sixes"].sum().rename("Sixes"),
    (deliveries.groupby(["inning","batsman"])["batsman_runs"].sum().rename("Highest Score").reset_index()
     .groupby(["batsman"])["Highest Score"].max().rename("Highest Score per Inning")),
     (deliveries[deliveries['Cumulative_batsman_runs']>=100].groupby(['batsman']).size().rename('Centuries')),
    (deliveries[deliveries['Cumulative_batsman_runs']>=50].groupby(['batsman']).size().rename('Half Centuries')),
    (deliveries[deliveries['cummulative_ball_count']==6].groupby(['batsman']).size().rename('Run Rate')),
    (deliveries[deliveries['Matches_played']>30].groupby(["match_id","batsman"])["Bounds"].sum().rename("Boundaries per match")
     .reset_index().groupby(["batsman"])["Boundaries per match"].mean().rename("Average Boundaries per match"))
     ],
       axis=1,
    )
    .fillna(0)
    .reset_index()
    .rename(columns={"index":"Batsman"})
                ) 
batsman_stats.sort_values(by='Total Runs Scored',ascending=False)[:5]

Batsman  Total Balls Faced  Balls at Non-Stricker's End  \
0    V Kohli             4211.0                       4071.0   
1   SK Raina             4044.0                       4173.0   
2  RG Sharma             3816.0                       3858.0   
6  DA Warner             3398.0                       3127.0   
3   S Dhawan             3776.0                       4090.0   

   Total Runs Scored  Maximum Runs in a Match  Fours  Sixes  \
0             5434.0                    113.0  482.0  191.0   
1             5415.0                    100.0  495.0  195.0   
2             4914.0                    109.0  431.0  194.0   
6             4741.0                    126.0  459.0  181.0   
3             4632.0                     98.0  526.0   96.0   

   Highest Score per Inning  Centuries  Half Centuries  Run Rate  \
0                    2876.0       13.0           482.0     151.0   
1                    3108.0        1.0           386.0     163.0   
2                    2837.0        9.0           351.0     155.0   
6                    2571.0       26.0           629.0     112.0   
3                    2690.0        0.0           364.0     139.0   

   Average Boundaries per match  
0                      4.000000  
1                      3.619048  
2                      3.447514  
6                      5.104000  
3                      3.917722

Batsman Strick Rate

In [52]:
batsman_stats['Strike Rate'] = (batsman_stats['Total Runs Scored']/batsman_stats['Total Balls Faced'])*100
batsman_stats

Batsman  Total Balls Faced  Balls at Non-Stricker's End  \
0          V Kohli             4211.0                       4071.0   
1         SK Raina             4044.0                       4173.0   
2        RG Sharma             3816.0                       3858.0   
3         S Dhawan             3776.0                       4090.0   
4        G Gambhir             3524.0                       3740.0   
..             ...                ...                          ...   
515    LH Ferguson                1.0                          3.0   
516          A Nel                0.0                          3.0   
517    C Ganapathy                0.0                          2.0   
518  BW Hilfenhaus                0.0                          1.0   
519   TP Sudhindra                0.0                          1.0   

     Total Runs Scored  Maximum Runs in a Match  Fours  Sixes  \
0               5434.0                    113.0  482.0  191.0   
1               5415.0                    100.0  495.0  195.0   
2               4914.0                    109.0  431.0  194.0   
3               4632.0                     98.0  526.0   96.0   
4               4223.0                     93.0  492.0   59.0   
..                 ...                      ...    ...    ...   
515                1.0                      1.0    0.0    0.0   
516                0.0                      0.0    0.0    0.0   
517                0.0                      0.0    0.0    0.0   
518                0.0                      0.0    0.0    0.0   
519                0.0                      0.0    0.0    0.0   

     Highest Score per Inning  Centuries  Half Centuries  Run Rate  \
0                      2876.0       13.0           482.0     151.0   
1                      3108.0        1.0           386.0     163.0   
2                      2837.0        9.0           351.0     155.0   
3                      2690.0        0.0           364.0     139.0   
4                      2464.0        0.0           397.0     131.0   
..                        ...        ...             ...       ...   
515                       1.0        0.0             0.0       0.0   
516                       0.0        0.0             0.0       0.0   
517                       0.0        0.0             0.0       0.0   
518                       0.0        0.0             0.0       0.0   
519                       0.0        0.0             0.0       0.0   

     Average Boundaries per match  Strike Rate  
0                        4.000000   129.042983  
1                        3.619048   133.902077  
2                        3.447514   128.773585  
3                        3.917722   122.669492  
4                        3.668874   119.835414  
..                            ...          ...  
515                      0.000000   100.000000  
516                      0.000000          NaN  
517                      0.000000          NaN  
518                      0.000000          NaN  
519                      0.000000          NaN  

[520 rows x 13 columns]

These are the top five "Super Strikers"

In [53]:
batsman_stats['Total Runs Scored'].sum()

223286.0

Total runs scored by all teams throughout 11 seasons is 2,23,286

In [54]:
batsman_stats.to_excel('C:/Users/Alexei/Desktop/Sparks Internship/Sports EDA.xlsx',sheet_name = 'Sheet1')

In [55]:
highest_individual_score = deliveries.groupby(['batsman','match_id'])['batsman_runs'].sum().sort_values(ascending=False)[:10]
highest_individual_score

batsman         match_id
CH Gayle        411         175
BB McCullum     60          158
AB de Villiers  562         133
RR Pant         7935        130
AB de Villiers  620         129
CH Gayle        372         128
M Vijay         206         127
DA Warner       36          126
V Sehwag        516         122
SR Watson       7953        121
Name: batsman_runs, dtype: int64

Chris Gayle scored the highest individual score throughout IPL

# Bowler Statistics

Top 5 "Tidy Bowlers"

In [56]:
bowler_stats = (pd.concat([
   deliveries.groupby(["bowler"])["over"].sum().rename("Overs"),
   deliveries.bowler.value_counts().rename("Total Balls Bowled"),
   deliveries.groupby(["bowler"])['Maiden Overs'].sum().rename("Maiden overs"),
   (deliveries[deliveries["dismissal_kind"].isin(["caught","bowled","lbw","stumped","caught and bowled","hit wicket"])]
   .groupby(["bowler"]).size().rename("Total Wickets")),
    deliveries.groupby(["bowler"])['No Ball'].sum().rename("No Balls"),
    deliveries.groupby(["bowler"])['Wide Ball'].sum().rename("Wide Balls"),
    (deliveries.groupby(["match_id","bowler"])["total_runs"].sum().rename("Total runs conceded").reset_index()
     .groupby(["bowler"])['Total runs conceded'].sum().rename("Total runs conceded per match")),
    (deliveries[deliveries["dismissal_kind"].isin(["caught","bowled","lbw","stumped","caught and bowled","hit wicket"])]
    .groupby(['match_id','bowler']).size().rename("Total Wickets").reset_index()
    .groupby(['bowler'])['Total Wickets'].mean().rename("Average Wickets taken per match"))
],
    axis=1,
)
.fillna(0)
.reset_index()
.rename(columns={"index":"Bowler"}))
bowler_stats.sort_values(by='Total Wickets',ascending=False)[:5]

Bowler  Overs  Total Balls Bowled  Maiden overs  Total Wickets  \
340       SL Malinga  32837                2974          1144          170.0   
7           A Mishra  36354                3172          1108          156.0   
126  Harbhajan Singh  32818                3451          1244          150.0   
264        PP Chawla  34907                3157          1100          149.0   
93          DJ Bravo  37145                2711           825          147.0   

     No Balls  Wide Balls  Total runs conceded per match  \
340        18         129                           3511   
7          20          50                           3850   
126         3          74                           4050   
264         1          31                           4153   
93          6         119                           3733   

     Average Wickets taken per match  
340                         1.868132  
7                           1.659574  
126                         1.612903  
264                         1.520408  
93                          1.651685

In [57]:
bowler_stats.to_excel('C:/Users/Alexei/Desktop/Sparks Internship/Sports EDA.xlsx',sheet_name = 'Sheet1')

In [58]:
bowler_stats['Total Wickets'].sum()

7968.0

Total wickets taken by bowlers throughout 11 seasons 

In [59]:
bowler_stats['Total Balls Bowled'].sum()

179078

Total balls bowled indicates the total number of deliveries

# Fielder Statistics

In [60]:
fielder_stats = (pd.concat([
    (deliveries[deliveries["dismissal_kind"].isin(["caught","caught and bowled"])].groupby(["fielder"])
    .size().rename("Catches")),
    (deliveries[deliveries['dismissal_kind']=='stumped'].groupby(['fielder']).size().rename("Stumpings")),
    (deliveries[deliveries['dismissal_kind']=='run out'].groupby(['fielder']).size().rename("Run Outs"))
],
axis=1,
)
.fillna(0)
.reset_index()
.rename(columns={"index":"Fielder"})
                )
fielder_stats.sort_values(by='Stumpings',ascending=False)[:5]

fielder  Catches  Stumpings  Run Outs
276      MS Dhoni     98.0       38.0      23.0
367    RV Uthappa     84.0       32.0       9.0
205    KD Karthik    109.0       29.0      14.0
477       WP Saha     60.0       18.0       4.0
21   AC Gilchrist     51.0       16.0       8.0

These are the top 5 fielders. MS Dhoni being in the lead.

In [61]:
fielder_stats['Run Outs'].sum()

848.0

Total run outs done throughout the IPL franchise

In [62]:
fielder_stats['Stumpings'].sum()

277.0

Total Stumpings 